src/v/cloud_storage/cache_service.cc

```diff
@@ -57,6 +57,38 @@ cache::cache(
   , _cnt(0)
   , _total_cleaned(0) {}
 
+ss::future<>
+cache::recursive_delete_empty_directory(const std::string_view& key) {
+    gate_guard guard{_gate};
+    std::filesystem::path path(key);
+
+    while (path != _cache_dir) {
+        try {
+            vassert(
+              std::string_view(path.native()).starts_with(_cache_dir.native()),
+              "Tried to clean up {}, which is outside of cache_dir {}.",
+              path.native(),
+              _cache_dir.native());
+
+            // ss::remove_file removes only empty directory
+            co_await ss::remove_file(path.native());
+        } catch (std::filesystem::filesystem_error& e) {
+            if (e.code() == std::errc::directory_not_empty) {
```

Comment:
you can use `co_await directory_walker::empty(path)` instead of relying on this particular behavior of remove_file.

Reply:
Even if I check that the directory is empty, another thread can concurrently create this directory before I delete it. The worst thing that can happen is that I'll delete some cache files. Also added re-try in `cache_service::put` so that it doesn't rely on directory to exist when creating new file.